src/main/java/com/artipie/asto/fs/FileStorage.java

```diff
@@ -72,27 +74,24 @@ public FileStorage(final Path path) {
     }
 
     @Override
-    public CompletableFuture<Collection<Key>> list(final String prefix) {
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
         return Single.fromCallable(
             () -> {
-                final String separator = FileSystems.getDefault().getSeparator();
-                if (!prefix.endsWith(separator)) {
-                    throw new IllegalArgumentException(
-                        String.format(
-                            "The prefix must end with '%s': \"%s\"",
-                            prefix,
-                            separator
-                        )
-                    );
+                final Path path = this.path(prefix);
+                final Collection<Key> keys;
+                if (Files.exists(path)) {
+                    final int dirnamelen = path.toString().length() - prefix.string().length();
+                    keys = Files.walk(path)
+                        .filter(Files::isRegularFile)
+                        .map(Path::toString)
+                        .map(p -> p.substring(dirnamelen))
+                        .map(s -> s.split(FileSystems.getDefault().getSeparator()))
+                        .map(Key.From::new)
+                        .sorted(Comparator.comparing(Key.From::string))
```

Comment:
@olegmoz could you please give me a hint why should we sort keys here? It's not so obvious for me. Or even better write a test instead of the explanation - that would be the best explanation =)

Reply:
@iakunin actually existing test checks order, because compares output with list, not a set. I added sorting, to make output independent from underlying OS. It happens that Unix and Windows do `Files.walk` in different order, maybe it also depends on file system. I think that it is a good idea, to guarantee order of keys regardless of OS/FS

Reply:
@olegmoz got it. Now it's much clearer for me, why it was done.